### ISSUE TITLE
Add (boolean expression) branch and function data to the collected coverage data

### ISSUE THREAD

**Issue by silkentrance** [benbria/coffee-coverage#25]
I am currently working on a PR that adds both boolean expression (if x == 1 and y ==2) branch coverage data as well as function invocation coverage data, similarly to what JSCover provides.

Differently from JSCover, the solution I have in mind would keep track of function names, e.g. "1"...N, and provide require source location information in the collected data. JSCover, from what I have figured out so far, just uses the function names (0...N) to collect data on their invocation, making it difficult for external tools, for example vows, to figure out on the whereabouts of the function in the original source.

Let me hear if this would be a useful addition to this tool. I know that it is rather "complicated" given the existing AST, but I believe that I am on the right track, so to say...

**Comment by jwalton** [comment]
Started doing some work on this today.  I like the visitor idea - it would be slick to implement JSCoverage style coverage and Istanbul coverage as just two different sets of visitors.  I have something in the works; will make a branch and push it here when I have something a bit more solid.  :)

**Comment by jwalton** [comment]
@silkentrance Have a look at https://github.com/benbria/coffee-coverage/tree/istanbul if you get a chance.  I have the basics of Istanbul support and branch/function/statement coverage here.

The /src/instrumentors folder has two "instrumentors"; one for the old JSCoverage style, and a new one for Istanbul.  There's a README.md that explains the interface an instrumentor has to implement.  Basically I took your visitor idea, but I simplified it a little; a visitor is now just a function defined on an instrumentor, so for example `Instabul.visitSwitch(node)` will get called once for every Switch node from coffee-script.  Even without Instanbul, this made the JSCoverage use case much simpler than it used to be; this was a really good idea.

I also made this `NodeWrapper` class, because I found I was passing around a `node` and a `nodeData` structure (which had the node's parent, and some other data) to every function in the world.  Once I combined these into one data structure, I realized it was a good place for functions like `insertBefore()`.  I have mixed feelings about NodeWrapper, because it kind of obscures the node from coffee-script... On the other hand, though, it might give us some isolation from changes in coffee-script, too.
